### ISSUE TITLE
Add leak tests for classes

### ISSUE THREAD

**Issue by sboof911** [pyccel/pyccel#1541]
### Relevant Discussion

https://github.com/pyccel/pyccel/pull/1508

### Describe the feature
After the PR https://github.com/pyccel/pyccel/pull/1508 is finished, now we need to add some tests in https://github.com/pyccel/pyccel/blob/devel/tests/ndarrays/leaks_check.py to check for any leaks.

### Test Code
Provide code which does not currently work but which should do when this issue is fixed:
```python
class Point(object):
    def __init__(self : 'Point'):
        self.x = np.array([1, 2, 3])
```

**Comment by verma-kartik** [comment]
Hey, @sboof911 @EmilyBourne, came across this issue. If no one is working on this, can I pick this up? Would like to work on Pyccel for some OSS contribution.

**Comment by EmilyBourne** [comment]
Hey @verma-kartik ,
Nobody is working on this currently. It would be great if you want to work on it :). The idea is to add a few slightly more complicated test cases with classes and run the translation through valgrind to make sure there are no memory leaks. Currently we do only this with ndarrays (numpy arrays) in the file mentioned by @sboof911.

Feel free to get in touch here or on discord if you need any more information about the issue

**Comment by verma-kartik** [comment]
@EmilyBourne I think I've set up the dev env. I have a workable editable copy on my system. I'll take few days to familiarise how Pyccel code works. 
also, you mentioned Valgrind, I'm on MacOs Ventura, valgrind is not supported on my system. Is that a hard requirement for you?

**Comment by EmilyBourne** [comment]
Huh! It's surprising that it's not supported at all on mac!

It's not a hard requirement but it is what we are using currently for the existing tests in the CI so it would be nice to keep the same workflow. That said this issue is just about writing the test so not having valgrind shouldn't actually be too problematic.

The existing test for arrays is here:
https://github.com/pyccel/pyccel/blob/devel/tests%2Fndarrays%2Fleaks_check.py
The CI script is here:
https://github.com/pyccel/pyccel/blob/6d7667bde21647e307c42296cf994458b93cb2aa/.github/actions/valgrind_run/action.yml#L8-L9
As you can see all it does is translate the code and run the leak test. For testing locally you could use any program (e.g. macos's `leak`). If you find leaks then that should be fine for opening associated issues.

**Comment by verma-kartik** [comment]
Yeah after Darwin, the valgrind peeps really don't wanna focus on it. 

Makes sense, `leak` would be enough for local. I'll get back on this to you in a few days.
